engine/src/flutter/runtime/runtime_controller_unittests.cc

```diff
@@ -0,0 +1,151 @@
+    this->actions.push_back(actions);
+  }
+
+  void SetSemanticsTreeEnabled(bool enabled) override {}
+
+  void HandlePlatformMessage(
+      std::unique_ptr<PlatformMessage> message) override {}
+
+  FontCollection& GetFontCollection() override { return font; }
+
+  std::shared_ptr<AssetManager> GetAssetManager() override { return nullptr; }
+
+  void OnRootIsolateCreated() override {};
+
+  void UpdateIsolateDescription(const std::string isolate_name,
+                                int64_t isolate_port) override {};
+
+  void SetNeedsReportTimings(bool value) override {};
+
+  std::unique_ptr<std::vector<std::string>> ComputePlatformResolvedLocale(
+      const std::vector<std::string>& supported_locale_data) override {
+    return nullptr;
+  }
+
+  void RequestDartDeferredLibrary(intptr_t loading_unit_id) override {}
+
+  std::weak_ptr<PlatformMessageHandler> GetPlatformMessageHandler()
+      const override {
+    return {};
+  }
+
+  void SendChannelUpdate(std::string name, bool listening) override {}
+
+  double GetScaledFontSize(double unscaled_font_size,
+                           int configuration_id) const override {
+    return 0.0;
+  }
+};
+
+class RuntimeControllerTester {
+ public:
+  explicit RuntimeControllerTester(UIDartState::Context& context)
+      : context_(context),
+        runtime_controller_(delegate_,
+                            nullptr,
+                            {},
+                            {},
+                            {},
+                            {},
+                            {},
+                            nullptr,
+                            context_) {}
+
+  void CanUpdateSemanticsWhenSetSemanticsTreeEnabled(SemanticsUpdate* update) {
+    ASSERT_TRUE(delegate_.updates.empty());
+    ASSERT_TRUE(delegate_.actions.empty());
+    runtime_controller_.UpdateSemantics(update);
+    // Semantics tree is not yet enabled.
+    ASSERT_TRUE(delegate_.updates.empty());
+    ASSERT_TRUE(delegate_.actions.empty());
+
+    runtime_controller_.SetSemanticsTreeEnabled(true);
+    runtime_controller_.UpdateSemantics(update);
+    ASSERT_FALSE(delegate_.updates.empty());
+    ASSERT_FALSE(delegate_.actions.empty());
+  }
+
+ private:
+  MockRuntimeDelegate delegate_;
+  UIDartState::Context& context_;
+  RuntimeController runtime_controller_;
+};
+
+TEST_F(RuntimeControllerTest, CanUpdateSemanticsWhenSetSemanticsTreeEnabled) {
+  // The code in this test is mostly setup code to get a SemanticsUpdate object.
+  // The real test is in RuntimeControllerTester::CanUpdateSemantics.
+  TaskRunners task_runners("test",                  // label
+                           GetCurrentTaskRunner(),  // platform
+                           CreateNewThread(),       // raster
+                           CreateNewThread(),       // ui
+                           CreateNewThread()        // io
+  );
+  UIDartState::Context context(task_runners);
+  std::shared_ptr<RuntimeControllerTester> tester =
```

Comment:
Fine to avoid repeating the type here:
```suggestion
  auto tester =
```